test/rules/use_function_declaration_instead_variable.dart

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N use_function_declaration_instead_variable`
+
+void main() {
+  thisIsGood() { // OK
+    print('this is good');
+  }
+
+  var bad1 = () { // LINT
+    print('this is bad');
+  };
+
+  var bad2 = bad1; // LINT
+
+  bad2 = bad1; // LINT
+
+  bad2 = () { // LINT
+    print('this is bad');
+  };
+
+  var bad3 = print; // LINT
```

Comment:
Again, I don't think preventing this is the intent of the guideline. Consider
```
var write = print;
if (args['writeToFile']) {
  write = (String s) {
    // write `s` to a file
  }
}
doWork(write);
```
If we define a local function named `write`, then we cannot re-assign it before passing it in somewhere. (Which means, by the way, that the lint should also not be produced when the variable is assigned another value even if the initializer contains a function literal.)

The point of the guideline is that if you're going to define a function, it's easier to read it if it looks like a function definition. The point isn't to prevent users from defining and using function-valued variables.